### PR TITLE
Testsuite: accept a list of tests to be run

### PR DESCRIPTION
**Pull request by neteler**
Developed by @mmacata, migrated here from https://github.com/GRASS-GIS/grass-ci/pull/4

For running all tests with multirunner wrapped in testsuite, we like to control which tests to run. Therefore this pull request adds arguments which
- allow to read from a list on filesystem to control which tests to run
- allow to write all testmodules to a list on filesystem, which can help to create above list

How to run:
```
(cd testsuite/examples/ ; \
  bash test_framework_GRASS_GIS_with_NC.sh test_framework_GRASS_GIS_with_NC.conf)
```

### CODE REVIEW

File: lib/python/gunittest/invoker.py [hunk 1]
```diff
@@ -240,7 +240,8 @@ def _run_test_module(self, module, results_dir, gisdbase, location):
             shutil.rmtree(mapset_dir)
 
     def run_in_location(self, gisdbase, location, location_type,
-                        results_dir):
+                        results_dir, test_list_in='False',
+                        test_list_out='False'):
```
Comment (wenzeslaus) [inline lib/python/gunittest/invoker.py]: `test_list_in=None`

File: lib/python/gunittest/invoker.py [hunk 2]
```diff
@@ -265,10 +266,31 @@ def run_in_location(self, gisdbase, location, location_type,
                                    universal_location_value=GrassTestLoader.universal_tests_value,
                                    import_modules=False)
 
+        if test_list_out != 'False':
```
Comment (wenzeslaus) [inline lib/python/gunittest/invoker.py]: `if test_list_out:`

File: lib/python/gunittest/invoker.py [hunk 3]
```diff
@@ -265,10 +266,31 @@ def run_in_location(self, gisdbase, location, location_type,
...
+            # overwrite old list
+            with open(test_list_out, 'w') as file:
+                file.write("# you can outcomment tests in the input list \n")
+            # append each module to list
+            for module in modules:
+                with open(test_list_out, 'a') as file:
+                    file.write(module.abs_file_path + "\n")
```
Comment (wenzeslaus) [inline lib/python/gunittest/invoker.py]: Operating on relative paths may make more sense here.

File: lib/python/gunittest/main.py [hunk 4]
```diff
@@ -135,13 +135,20 @@ def main():
     parser.add_argument('--output', dest='output', action='store',
                         default='testreport',
                         help='Output directory')
+    parser.add_argument('--test_list_in',
```
Comment (wenzeslaus) [inline lib/python/gunittest/main.py]: `--skip`, `--select`, `--include`, `--whitelist`, ...?

File: lib/python/gunittest/main.py [hunk 5]
```diff
@@ -135,13 +135,20 @@ def main():
...
+                        help='File to read tests from list', default='False')
+    parser.add_argument('--test_list_out',
+                        help='File to write all test paths to list', default='False')
```
Comment (wenzeslaus) [inline lib/python/gunittest/main.py]: Needs rewording.

File: lib/python/gunittest/main.py [hunk 6]
```diff
@@ -135,13 +135,20 @@ def main():
...
+
     args = parser.parse_args()
     gisdbase = args.gisdbase
     if gisdbase is None:
         # here we already rely on being in GRASS session
         gisdbase = gcore.gisenv()['GISDBASE']
     location = args.location
     location_type = args.location_type
+    test_list_in = args.test_list_in
+    test_list_out = args.test_list_out
```
Comment (wenzeslaus) [inline lib/python/gunittest/main.py]: Not needed given the code below.

File: testsuite/examples/test.list [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# you can outcomment tests in the input list
```
Comment (wenzeslaus) [inline testsuite/examples/test.list]: comment out

File: testsuite/examples/test.list [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# you can outcomment tests in the input list 
+/home/user/grass_master/vector/v.in.pdal/testsuite/test_v_in_pdal_basic.py
```
Comment (wenzeslaus) [inline testsuite/examples/test.list]: Here the relative paths would be nice, so the file is actually usable.

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf [hunk 9]
```diff
@@ -27,3 +27,15 @@ REPORTS="testreports"
 PUBLISH="no"
 # upload WWW dir on server for report publication (not used for local tests)
 SERVERDIR="/var/www/html/grassgistestreports"
+
+# read user defined module tests list from file
+# may be no|yes
+READ_MODULE_TEST_LIST="no"
+# file with list of tests to be run
+TEST_LIST_IN="${GRASSSRC}/testsuite/examples/test.list"
```
Comment (wenzeslaus) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: I would suggest a more standard extension like `.txt`.

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.sh [hunk 10]
```diff
@@ -66,6 +66,7 @@ set -e  # fail fast
 
 # computer architecture:
 ARCH=`${GRASSBIN} --config arch`
+export GISBASE=`${GRASSBIN} --config path`
```
Comment (wenzeslaus) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.sh]: Related to this PR?

File: lib/python/gunittest/multirunner.py [hunk 11]
```diff
@@ -147,7 +153,9 @@ def main():
                               '-m', 'grass.gunittest.main',
                               '--grassdata', gisdb, '--location', location,
                               '--location-type', location_type,
-                              '--output', absreport],
+                              '--output', absreport,
+                              '--test_list_in', test_list_in,
+                              '--test_list_out', test_list_out],
```
Comment (wenzeslaus) [inline lib/python/gunittest/multirunner.py]: multirunner and output file does not make sense. If nothing else, the current implementation just overwrites the file.

File: lib/python/gunittest/invoker.py [hunk 12]
```diff
@@ -265,10 +266,31 @@ def run_in_location(self, gisdbase, location, location_type,
...
-        for module in modules:
+
+        test_modules = modules
+
+        if test_list_in != 'False':
+            test_modules = []
+            for module in modules:
+                with open(test_list_in, 'r') as testwhitelist:
+                    for line in testwhitelist:
+                        if str(module.abs_file_path) == str(line.strip()):
```
Comment (wenzeslaus) [inline lib/python/gunittest/invoker.py]: The str conversions don't look safe.

File: lib/python/gunittest/main.py [hunk 13]
```diff
@@ -176,7 +182,9 @@ def main():
         gisdbase=gisdbase,
         location=location,
         location_type=location_type,
-        results_dir=results_dir
+        results_dir=results_dir,
+        test_list_in=args.test_list_in,
+        test_list_out=args.test_list_out)
```
Comment (neteler) [inline lib/python/gunittest/main.py]:
```suggestion
        test_list_out=args.test_list_out
```

**Comment by neteler** [comment]
Please feel free to direct commit into this PR, thanks

**Comment by neteler** [comment]
I'm still interested in this PR. Just I struggle to rebase it...

**Comment by wenzeslaus** [comment]
See my review. It might be better to start over with in a new PR ideally taking a different approach. I suggest to close this PR, open an issue if needed.

To add to my review:

6. Python unittest can select specific tests (classes or functions) and that already works with gunittest when running individual test files. Taking that mechanism, possibly extending it for files or testsuite directories and adding reverse selection as an alternative to introducing another system is something to consider.

**Comment by neteler** [comment]
> See my review. It might be better to start over with in a new PR ideally taking a different approach. I suggest to close this PR, open an issue if needed.

Agreed and done so: #996

**Pull request closed without merging**